File: 958.check-completeness-of-a-binary-tree.py
#
# @lc app=leetcode id=958 lang=python3
#
# [958] Check Completeness of a Binary Tree
#

# @lc code=start

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
from typing import Optional


class Solution:
    '''
    Simple. Just breadth-first search from root. When we insert into root, we insert the node count as well. For example, in the first node, we have i = 1. Then it's left child has count of 2*i and right child has count of 2*i+1    
    '''

    def isCompleteTree(self, root: Optional[TreeNode]) -> bool:
        dq, total = deque(), 0
        dq.append([root, 1])
        while dq:
            for i in range(len(dq)):
                node, count = dq.popleft()
                total += 1
                if count != total:
                    return False
                if node.left:
                    dq.append([node.left, 2*count])
                if node.right:
                    dq.append([node.right, 2*count + 1])
        return True
# @lc code=end

File: 958.check-completeness-of-a-binary-tree.cs
/*
 * @lc app=leetcode id=958 lang=csharp
 *
 * [958] Check Completeness of a Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 using System.Collections.Generic;
 using System;
public class Solution 
{
    public bool IsCompleteTree(TreeNode root) 
    {
        Queue<object[]> queue = new Queue<object[]>();
        int total = 0;
        queue.Enqueue(new object[2] {root, 1});

        while (queue.Count > 0)
        {
            int n = queue.Count;
            for (int i = 0; i < n; i++)
            {
                object[] item = queue.Dequeue();
                total++;
                if (Convert.ToInt32(item[1]) != total)
                {
                    return false;
                }
                if ((item[0] as TreeNode).left != null)
                {
                    queue.Enqueue(new object[2]{(item[0] as TreeNode)?.left, 2*Convert.ToInt32(item[1])});
                }
                if ((item[0] as TreeNode).right != null)
                {
                    queue.Enqueue(new object[2]{(item[0] as TreeNode)?.right, 2*Convert.ToInt32(item[1]) + 1});
                }
            }
        }
        return true;
    }
}
// @lc code=end

